### PR TITLE
Add helper functions for setting dative bonds to transition metals

### PR DESCRIPTION
**Pull request by michaelmaser**
PTAL. There may be edge cases I'm not thinking of for carbon-metal or other X-M bonds that this wouldn't handle properly. E.g., if carbon is both metal-bound and additionally anionic, the C-M bond would be converted (which may not be desired) since the total valence + formal charge != 4. I don't know of any examples of this, but let me know if there are any other things like this you think of. I will send a notebook of tests for NHC ligands that demonstrate this handling.

### CODE REVIEW

**Comment by skearnes** [comment]
I think I forgot about this; sorry. Could you take a look at the failing test?

**Comment by michaelmaser** [comment]
Looks like may be a yapf error. 

From log: 

>2020-11-24T21:59:07.5045751Z Successfully installed yapf-0.30.0
>2020-11-24T21:59:10.8625742Z ##[error]Process completed with exit code 1.
>2020-11-24T21:59:10.8734066Z Post job cleanup.

From an old yapf GitHub issue: 

>A return code of `0` indicates that nothing was formatted and the program didn't have an internal failure.
>A return code of `1` indicates that the program had an internal failure. It's this way because Python apparently returns `1` in such cases.
>A return code of 2 indicates that some files were changed.

Any thoughts on what this may be?

**Comment by skearnes** [comment]
> Looks like may be a yapf error.
> 
> From log:
> 
> > 2020-11-24T21:59:07.5045751Z Successfully installed yapf-0.30.0
> > 2020-11-24T21:59:10.8625742Z ##[error]Process completed with exit code 1.
> > 2020-11-24T21:59:10.8734066Z Post job cleanup.
> 
> From an old yapf GitHub issue:
> 
> > A return code of `0` indicates that nothing was formatted and the program didn't have an internal failure.
> > A return code of `1` indicates that the program had an internal failure. It's this way because Python apparently returns `1` in such cases.
> > A return code of 2 indicates that some files were changed.
> 
> Any thoughts on what this may be?

Did you run the `./format.sh` script on your code? You probably have something that yapf doesn't like.

**Comment by skearnes** [comment]
I ran `./format.sh` locally and there are a bunch of changes in `message_helpers.py`; committing those should fix the problem.

File: ord_schema/message_helpers.py
Comment (skearnes) [inline ord_schema/message_helpers.py:533]: This can be expressed more succinctly:

`return 22 <= n <= 29 or 40 <= n <= 47 or 72 <= n <= 79`

File: ord_schema/message_helpers.py
Comment (skearnes) [inline ord_schema/message_helpers.py:552]: Please make the first line standalone and descriptive (rather than imperative); see https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

```python
"""Converts metal-ligand bonds to dative.

Replaces some ...
"""
```

File: ord_schema/message_helpers.py
Comment (skearnes) [inline ord_schema/message_helpers.py:562]: Default is `<the atomic numbers for N and P>`.

Maybe use the atom symbols here instead?

File: ord_schema/message_helpers.py
Comment (skearnes) [inline ord_schema/message_helpers.py:579]: Please use parentheses instead of line breaks.

File: ord_schema/message_helpers.py
Comment (skearnes) [inline ord_schema/message_helpers.py:581]: Can you just update the type of the bond instead of deleting it and adding a new one?

**Review by skearnes — CHANGES_REQUESTED** [review]
Please add a test for these methods in `message_helpers_test.py`.

**Comment by michaelmaser** [comment]
> I ran `./format.sh` locally and there are a bunch of changes in `message_helpers.py`; committing those should fix the problem.

Fixed! Thanks.

File: ord_schema/message_helpers.py
Comment (michaelmaser) [reply ord_schema/message_helpers.py]: Changed!

File: ord_schema/message_helpers.py
Comment (michaelmaser) [reply ord_schema/message_helpers.py]: Done, thanks.

File: ord_schema/message_helpers.py
Comment (michaelmaser) [reply ord_schema/message_helpers.py:564]: Done. Good idea, thanks!

File: ord_schema/message_helpers.py
Comment (michaelmaser) [reply ord_schema/message_helpers.py]: Done

File: ord_schema/message_helpers.py
Comment (michaelmaser) [reply ord_schema/message_helpers.py]: I believe this can be done with `ReplaceBond`, but requires creating a new `Bond` object and getting the bond index, etc. Can be done, I only worry about the direction (`BondDir`) of the dative bond being correct. Can check with `GetBondDir` or `GetStartAtom`, but ends up being a bit lengthier than what's there now. Is there another reason to use `ReplaceBond` over this?

**Comment by michaelmaser** [comment]
Maybe hold off on merging for now. Want to do some testing to make sure changes are working. 

pylint is unhappy about whitespace. Will fix and test functions tomorrow.

File: ord_schema/message_helpers.py
Comment (connorcoley) [reply ord_schema/message_helpers.py]: I think remove/add is a fine solution -- the code looks nice

Comment (skearnes) [reply ord_schema/message_helpers.py]: Ah I didn't appreciate the control over the direction. SGTM.

**Comment by skearnes** [comment]
LGTM; I'll approve after some tests are added.

**Comment by michaelmaser** [comment]
Tests added!

File: ord_schema/message_helpers_test.py
Comment (skearnes) [inline ord_schema/message_helpers_test.py:395]: These look pretty good, but a couple of tips on writing tests:

* In general, try to make the test as simple as possible. It should be trivial for a reviewer to verify that the test is correct. For instance, instead of creating a `Compound` message with a complicated SMILES identifier, you could just do `self.assertFalse(message_helpers.has_transition_metal(Chem.MolFromSmiles('P')))`. AFAICT there's no reason to create the proto wrapper or use such a complicated structure for what you're testing here.
* Prefer parameterized tests over listing multiple test cases in the same method. See, for example, https://github.com/Open-Reaction-Database/ord-schema/blob/main/ord_schema/message_helpers_test.py#L95.

Comment (michaelmaser) [reply ord_schema/message_helpers_test.py:395]: >AFAICT there's no reason to create the proto wrapper or use such a complicated structure for what you're testing here.

I see, thanks. The only reason I added the wrapper was to make sure the method worked within the proto. I had already checked that the function worked in RDKit alone, but I see that's what we're going for in the test here too. Will update.

>Prefer parameterized tests over listing multiple test cases in the same method.

Sounds great. Will work on these now.

Comment (michaelmaser) [reply ord_schema/message_helpers_test.py]: Simplified/parameterized tests. Plan to add a couple more simple ones to `test_set_dative_bonds` tonight or tomorrow

File: ord_schema/message_helpers_test.py
Comment (skearnes) [inline ord_schema/message_helpers_test.py:391]: Much clearer; thanks!

**Review by skearnes — APPROVED** [review]

**Pull request merged**
